Memory Leak in Zed Editor's Built-in Terminal When Running yes $(echo -en "\a")

### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

**Description:**
When using Zed Editor's built-in terminal, running the command `yes $(echo -en "\a")` causes the editor to rapidly consume increasing amounts of memory (approximately 0.1GB every 1-2 seconds). Even after closing the terminal, the memory is not released.

**Steps to Reproduce:**
1. Open Zed Editor.
2. Launch the built-in terminal within Zed.
3. Run the command: `yes $(echo -en "\a")`.
4. Observe the memory usage of Zed Editor (e.g., via system monitor or task manager).
5. Notice that memory usage increases rapidly (0.1GB every 1-2 seconds).
6. Close the terminal tab or window.
7. Observe that the memory previously consumed by the terminal is not released.

### Zed Version and System Specs

Zed: v0.168.2 (Zed)
OS: Linux Wayland arch unknown
Memory: 30.7 GiB
Architecture: x86_64
GPU: NVIDIA GeForce RTX 3050 Laptop GPU || NVIDIA || 565.77

### If applicable, add screenshots or screencasts of the incorrect state / behavior

![eats RAM](https://github.com/user-attachments/assets/f3898d00-ebed-41c9-9b2d-055113b833e2)

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary>

<!-- Click below this line and paste or drag-and-drop your log-->
```
2025-01-12T02:28:45.89080675+08:00 [INFO] ========== starting zed ==========
2025-01-12T02:28:45.985662964+08:00 [INFO] Presentation support: "VK_KHR_wayland_surface"
2025-01-12T02:28:45.98579252+08:00 [INFO] Presentation support: "VK_KHR_xcb_surface"
2025-01-12T02:28:45.98589358+08:00 [INFO] Presentation support: "VK_KHR_xlib_surface"
2025-01-12T02:28:45.985993733+08:00 [INFO] Enabling Vulkan Portability
2025-01-12T02:28:45.986082082+08:00 [INFO] Enabling color space support
2025-01-12T02:28:45.999699479+08:00 [INFO] Adapter: "NVIDIA GeForce RTX 3050 Laptop GPU"
2025-01-12T02:28:45.999928419+08:00 [INFO] Ray tracing is supported
2025-01-12T02:28:46.362928816+08:00 [INFO] perform;
2025-01-12T02:28:46.363092524+08:00 [INFO] authenticate;
2025-01-12T02:28:46.363224105+08:00 [INFO] Opening main db
2025-01-12T02:28:46.363334734+08:00 [INFO] read_command;
2025-01-12T02:28:46.363474626+08:00 [INFO] send_secondary_commands;
2025-01-12T02:28:46.363878868+08:00 [INFO] new;
2025-01-12T02:28:46.36399711+08:00 [INFO] socket reader;
2025-01-12T02:28:46.364675549+08:00 [INFO] keep_updated;
2025-01-12T02:28:46.402201263+08:00 [INFO] Opening main db
2025-01-12T02:28:46.404934786+08:00 [INFO] Using git binary path: None
2025-01-12T02:28:46.434566676+08:00 [ERROR] theme not found: KTRZ Monokai
2025-01-12T02:28:46.435981803+08:00 [INFO] No prompt template overrides directory found at /home/orreo/.var/app/dev.zed.Zed/data/zed/prompt_overrides. Using built-in prompts.
2025-01-12T02:28:46.436537181+08:00 [INFO] No prompt template overrides directory found at /home/orreo/.var/app/dev.zed.Zed/data/zed/prompt_overrides. Using built-in prompts.
2025-01-12T02:28:46.436988846+08:00 [INFO] extensions updated. loading 5, reloading 0, unloading 0
2025-01-12T02:28:46.439033453+08:00 [INFO] activate is not implemented on Linux, ignoring the call
2025-01-12T02:28:46.440254979+08:00 [INFO] Opening main db
2025-01-12T02:28:46.440771596+08:00 [INFO] perform;
2025-01-12T02:28:46.440905132+08:00 [INFO] authenticate;
2025-01-12T02:28:46.441053336+08:00 [INFO] read_command;
2025-01-12T02:28:46.441594816+08:00 [INFO] socket reader;
2025-01-12T02:28:46.446405841+08:00 [INFO] building git repository, `.git` path in the worktree: ".git"
2025-01-12T02:28:46.446901156+08:00 [INFO] new;
2025-01-12T02:28:46.44768792+08:00 [INFO] keep_updated;
2025-01-12T02:28:46.465448657+08:00 [INFO] Using surface present mode MAILBOX
2025-01-12T02:28:46.472935656+08:00 [INFO] Creating a descriptor pool for at most 16 sets
2025-01-12T02:28:46.473057738+08:00 [INFO] Creating a descriptor pool for at most 16 sets
2025-01-12T02:28:46.473156564+08:00 [INFO] Initializing Blade pipelines for surface SurfaceInfo { format: Bgra8UnormSrgb, alpha: PreMultiplied }
2025-01-12T02:28:46.484850948+08:00 [INFO] Using GPU: GpuSpecs { is_software_emulated: false, device_name: "NVIDIA GeForce RTX 3050 Laptop GPU", driver_name: "NVIDIA", driver_info: "565.77" }
2025-01-12T02:28:46.485449279+08:00 [INFO] set status on client 0: Authenticating
2025-01-12T02:28:46.485934956+08:00 [INFO] perform;
2025-01-12T02:28:46.486076874+08:00 [INFO] authenticate;
2025-01-12T02:28:46.486225077+08:00 [INFO] read_command;
2025-01-12T02:28:46.486710056+08:00 [INFO] socket reader;
2025-01-12T02:28:46.488278484+08:00 [INFO] Opening main db
2025-01-12T02:28:46.492449971+08:00 [INFO] new;
2025-01-12T02:28:46.494196984+08:00 [INFO] keep_updated;
2025-01-12T02:28:46.494605277+08:00 [INFO] set environment variables from shell:/bin/zsh, path:/home/orreo/bin:/usr/local/bin:/usr/local/sbin:/home/orreo/.guix-profile/bin:/home/orreo/.config/guix/current/bin:/home/orreo/.volta/bin:/usr/bin:/home/orreo/.local/share/flatpak/exports/bin:/var/lib/flatpak/exports/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/home/orreo/go/bin:/home/orreo/.cargo/bin/:/home/orreo/go/bin:/home/orreo/.cargo/bin/
2025-01-12T02:28:46.620434686+08:00 [INFO] set status on client 285599: Connecting
2025-01-12T02:28:46.632245565+08:00 [INFO] Using surface present mode MAILBOX
2025-01-12T02:28:46.632404035+08:00 [INFO] Retaining current format: Bgra8UnormSrgb
2025-01-12T02:28:46.651224477+08:00 [INFO] Using surface present mode MAILBOX
2025-01-12T02:28:46.651341601+08:00 [INFO] Retaining current format: Bgra8UnormSrgb
2025-01-12T02:28:46.67903637+08:00 [INFO] using project environment variables from CLI. PATH="/home/orreo/bin:/usr/local/bin:/usr/local/sbin:/home/orreo/.guix-profile/bin:/home/orreo/.config/guix/current/bin:/home/orreo/.volta/bin:/usr/bin:/home/orreo/.local/share/flatpak/exports/bin:/var/lib/flatpak/exports/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/home/orreo/go/bin:/home/orreo/.cargo/bin/"
2025-01-12T02:28:46.680664303+08:00 [INFO] perform;
2025-01-12T02:28:46.680820748+08:00 [INFO] authenticate;
2025-01-12T02:28:46.680971395+08:00 [INFO] read_command;
2025-01-12T02:28:46.681628743+08:00 [INFO] socket reader;
2025-01-12T02:28:46.68757036+08:00 [INFO] new;
2025-01-12T02:28:46.688367739+08:00 [INFO] keep_updated;
2025-01-12T02:28:48.47325839+08:00 [WARN] unhandled message: Metadata(SupermavenMetadataMessage { dust_strings: None })
2025-01-12T02:28:48.764545084+08:00 [INFO] Using surface present mode MAILBOX
2025-01-12T02:28:48.76470551+08:00 [INFO] Retaining current format: Bgra8UnormSrgb
2025-01-12T02:28:50.036503048+08:00 [INFO] connected to rpc endpoint https://collab.zed.dev/rpc
2025-01-12T02:28:53.120745507+08:00 [INFO] add_connection;
2025-01-12T02:28:53.132965238+08:00 [INFO] set status on client 285599: Connected { peer_id: PeerId { owner_id: 633, id: 148913 }, connection_id: ConnectionId { owner_id: 0, id: 0 } }
2025-01-12T02:28:54.694285064+08:00 [ERROR] error sending request for url (https://api.zed.dev/telemetry/events?)

Caused by:
    0: client error (Connect)
    1: tls handshake eof
2025-01-12T02:30:10.705772812+08:00 [INFO] Using surface present mode MAILBOX
2025-01-12T02:30:10.705912495+08:00 [INFO] Retaining current format: Bgra8UnormSrgb
```
<!-- Click above this line and paste or drag-and-drop your log--></details>

--- a/crates/repl/src/outputs/plain.rs
+++ b/crates/repl/src/outputs/plain.rs
@@ -16,6 +16,7 @@
 //!
 
 use alacritty_terminal::{
+    event::VoidListener,
     grid::Dimensions as _,
     index::{Column, Line, Point},
     term::Config,
@@ -24,8 +25,6 @@ use alacritty_terminal::{
 use gpui::{canvas, size, ClipboardItem, FontStyle, Model, TextStyle, WhiteSpace};
 use language::Buffer;
 use settings::Settings as _;
-use std::mem;
-use terminal::ZedListener;
 use terminal_view::terminal_element::TerminalElement;
 use theme::ThemeSettings;
 use ui::{prelude::*, IntoElement};
@@ -50,7 +49,7 @@ pub struct TerminalOutput {
     /// ANSI escape sequence processor for parsing input text.
     parser: Processor,
     /// Alacritty terminal instance that manages the terminal state and content.
-    handler: alacritty_terminal::Term<ZedListener>,
+    handler: alacritty_terminal::Term<VoidListener>,
 }
 
 const DEFAULT_NUM_LINES: usize = 32;
@@ -124,14 +123,9 @@ impl TerminalOutput {
     /// and sets up the necessary components for handling terminal events and rendering.
     ///
     pub fn new(cx: &mut WindowContext) -> Self {
-        let (events_tx, events_rx) = futures::channel::mpsc::unbounded();
-        let term = alacritty_terminal::Term::new(
-            Config::default(),
-            &terminal_size(cx),
-            terminal::ZedListener(events_tx.clone()),
-        );
+        let term =
+            alacritty_terminal::Term::new(Config::default(), &terminal_size(cx), VoidListener);
 
-        mem::forget(events_rx);
         Self {
             parser: Processor::new(),
             handler: term,

--- a/crates/terminal/src/terminal.rs
+++ b/crates/terminal/src/terminal.rs
@@ -28,7 +28,7 @@ use anyhow::{bail, Result};
 
 use futures::{
     channel::mpsc::{unbounded, UnboundedReceiver, UnboundedSender},
-    FutureExt,
+    FutureExt, SinkExt,
 };
 
 use mappings::mouse::{
@@ -53,7 +53,7 @@ use std::{
     ops::{Deref, Index, RangeInclusive},
     path::PathBuf,
     sync::Arc,
-    time::Duration,
+    time::{Duration, Instant},
 };
 use thiserror::Error;
 
@@ -482,66 +482,141 @@ impl TerminalBuilder {
         })
     }
 
-    pub fn subscribe(mut self, cx: &ModelContext<Terminal>) -> Terminal {
-        //Event loop
-        cx.spawn(|terminal, mut cx| async move {
-            while let Some(event) = self.events_rx.next().await {
-                terminal.update(&mut cx, |terminal, cx| {
-                    //Process the first event immediately for lowered latency
-                    terminal.process_event(&event, cx);
-                })?;
-
-                'outer: loop {
-                    let mut events = Vec::new();
-                    let mut timer = cx
-                        .background_executor()
-                        .timer(Duration::from_millis(4))
-                        .fuse();
-                    let mut wakeup = false;
-                    loop {
-                        futures::select_biased! {
-                            _ = timer => break,
-                            event = self.events_rx.next() => {
-                                if let Some(event) = event {
-                                    if matches!(event, AlacTermEvent::Wakeup) {
-                                        wakeup = true;
-                                    } else {
-                                        events.push(event);
-                                    }
-
-                                    if events.len() > 100 {
+    pub fn subscribe(self, cx: &ModelContext<Terminal>) -> Terminal {
+        // Accumulate the effects of events on a background thread in order to keep up with the
+        // events from alacritty even when it is emitting events rapidly.
+        let (mut accumulated_events_tx, mut accumulated_events_rx) = unbounded();
+        let mut events_rx = self.events_rx;
+        let background_executor = cx.background_executor().clone();
+        cx.background_executor()
+            .spawn(async move {
+                while let Some(event) = events_rx.next().await {
+                    // Process the first event immediately to reduce latency
+                    accumulated_events_tx
+                        .feed(EventOrAccumulator::Event(event))
+                        .await?;
+                    'outer: loop {
+                        let start_time = Instant::now();
+                        let mut timer = background_executor.timer(Duration::from_millis(4)).fuse();
+                        let mut event_accumulator = EventAccumulator::new();
+                        loop {
+                            futures::select_biased! {
+                                // Events are no longer coming in at a high rate, so go back to just
+                                // awaiting the next event.
+                                _ = timer => break 'outer,
+                                event = events_rx.next() => {
+                                    let Some(event) = event else {
+                                        break;
+                                    };
+                                    event_accumulator.add(event);
+                                    if event_accumulator.events.len() > 100 {
                                         break;
                                     }
-                                } else {
-                                    break;
-                                }
-                            },
+                                    let elapsed = Instant::now().duration_since(start_time);
+                                    if elapsed > Duration::from_millis(20) {
+                                        break;
+                                    }
+                                },
+                            }
                         }
+                        accumulated_events_tx
+                            .feed(EventOrAccumulator::Accumulator(event_accumulator))
+                            .await?;
                     }
-
-                    if events.is_empty() && !wakeup {
-                        smol::future::yield_now().await;
-                        break 'outer;
-                    }
-
-                    terminal.update(&mut cx, |this, cx| {
-                        if wakeup {
-                            this.process_event(&AlacTermEvent::Wakeup, cx);
-                        }
-
-                        for event in events {
-                            this.process_event(&event, cx);
-                        }
-                    })?;
-                    smol::future::yield_now().await;
                 }
-            }
+                anyhow::Ok(())
+            })
+            .detach();
 
+        // On the foreground thread, process the accumulated effects of events.
+        cx.spawn(|terminal, mut cx| async move {
+            while let Some(event_or_accumulator) = accumulated_events_rx.next().await {
+                terminal.update(&mut cx, |terminal, cx| {
+                    event_or_accumulator.process_events(terminal, cx)
+                })?;
+            }
             anyhow::Ok(())
         })
         .detach();
 
         self.terminal
+    }
+}
+
+enum EventOrAccumulator {
+    Event(AlacTermEvent),
+    Accumulator(EventAccumulator),
+}
+
+impl EventOrAccumulator {
+    fn process_events(self, terminal: &mut Terminal, cx: &mut ModelContext<Terminal>) {
+        match self {
+            EventOrAccumulator::Event(event) => terminal.process_event(event, cx),
+            EventOrAccumulator::Accumulator(accumulator) => {
+                accumulator.process_events(terminal, cx)
+            }
+        }
+    }
+}
+
+struct EventAccumulator {
+    wakeup: bool,
+    cursor_blinking_changed: bool,
+    bell: bool,
+    title: Option<String>,
+    /// Events that can't be deduplicated.
+    events: Vec<AlacTermEvent>,
+}
+
+impl EventAccumulator {
+    fn new() -> Self {
+        EventAccumulator {
+            wakeup: false,
+            cursor_blinking_changed: false,
+            bell: false,
+            title: None,
+            events: Vec::new(),
+        }
+    }
+
+    fn add(&mut self, event: AlacTermEvent) {
+        match event {
+            // Events that can have their effects deduplicated.
+            AlacTermEvent::Title(title) => self.title = Some(title),
+            AlacTermEvent::ResetTitle => self.title = Some(String::new()),
+            AlacTermEvent::CursorBlinkingChange => self.cursor_blinking_changed = true,
+            AlacTermEvent::Wakeup => self.wakeup = true,
+            AlacTermEvent::Bell => self.bell = true,
+            // Events that have handlers involving writing text to the terminal or interacting with
+            // clipboard, and so must be kept in order.
+            AlacTermEvent::ClipboardStore(_, _) => self.events.push(event),
+            AlacTermEvent::ClipboardLoad(_, _) => self.events.push(event),
+            AlacTermEvent::PtyWrite(_) => self.events.push(event),
+            AlacTermEvent::TextAreaSizeRequest(_) => self.events.push(event),
+            AlacTermEvent::ColorRequest(_, _) => self.events.push(event),
+            AlacTermEvent::Exit => self.events.push(event),
+            AlacTermEvent::ChildExit(_) => self.events.push(event),
+            // Handled in render so no need to handle here.
+            AlacTermEvent::MouseCursorDirty => {}
+        }
+    }
+
+    fn process_events(self, terminal: &mut Terminal, cx: &mut ModelContext<Terminal>) {
+        if self.wakeup {
+            terminal.process_event(AlacTermEvent::Wakeup, cx);
+        }
+        if self.cursor_blinking_changed {
+            terminal.process_event(AlacTermEvent::CursorBlinkingChange, cx);
+        }
+        if self.bell {
+            terminal.process_event(AlacTermEvent::Bell, cx);
+        }
+        if let Some(title) = self.title {
+            terminal.process_event(AlacTermEvent::Title(title), cx);
+        }
+        for event in self.events {
+            terminal.process_event(event, cx);
+        }
     }
 }
 
@@ -674,7 +749,7 @@ impl TaskStatus {
 }
 
 impl Terminal {
-    fn process_event(&mut self, event: &AlacTermEvent, cx: &mut ModelContext<Self>) {
+    fn process_event(&mut self, event: AlacTermEvent, cx: &mut ModelContext<Self>) {
         match event {
             AlacTermEvent::Title(title) => {
                 self.breadcrumb_text = title.to_string();
@@ -728,13 +803,12 @@ impl Terminal {
                 // Instead of locking, we could store the colors in `self.last_content`. But then
                 // we might respond with out of date value if a "set color" sequence is immediately
                 // followed by a color request sequence.
-                let color = self.term.lock().colors()[*index].unwrap_or_else(|| {
-                    to_alac_rgb(get_color_at_index(*index, cx.theme().as_ref()))
-                });
+                let color = self.term.lock().colors()[index]
+                    .unwrap_or_else(|| to_alac_rgb(get_color_at_index(index, cx.theme().as_ref())));
                 self.write_to_pty(format(color));
             }
             AlacTermEvent::ChildExit(error_code) => {
-                self.register_task_finished(Some(*error_code), cx);
+                self.register_task_finished(Some(error_code), cx);
             }
         }
     }

